File: game.py
import os
from engine import Board, SHIPS, SHOT_PATTERNS, BOARD_SIZE
from ai import AI

def clear_screen():
    os.system('cls' if os.name == 'nt' else 'clear')

def show_patterns():
    """Displays visual hints for all ship shot patterns."""
    print("\n--- SHOT PATTERN REFERENCE ---")
    print("O = Origin (Your target), X = Hit area, . = Empty")
    
    for ship, pattern in SHOT_PATTERNS.items():
        print(f"\n{ship}:")
        # Find bounds of the pattern
        min_x = min(dx for dx, dy in pattern + [(0,0)])
        max_x = max(dx for dx, dy in pattern + [(0,0)])
        min_y = min(dy for dx, dy in pattern + [(0,0)])
        max_y = max(dy for dx, dy in pattern + [(0,0)])
        
        # Add a bit of padding
        for y in range(min_y, max_y + 1):
            row = []
            for x in range(min_x, max_x + 1):
                if (x, y) == (0, 0):
                    row.append("O")
                elif (x, y) in pattern:
                    row.append("X")
                else:
                    row.append(".")
            print(" " + " ".join(row))
    print("------------------------------")

class Game:
    def __init__(self):
        self.player_board = Board()
        self.player_board.place_randomly(SHIPS)
        
        self.ai_board = Board()
        self.ai_board.place_randomly(SHIPS)
        
        self.ai_agent = AI(difficulty='experiment')

    def render_board(self, reveal_enemy=False):
        """Displays the player's board and the tracking board for enemy shots."""
        # Header labels centered above the grids
        print("\n        PLAYER BOARD                     ENEMY BOARD")
        print("        (Your Ships)                    (Target Area)")
        
        header = "     " + " ".join([str(i) for i in range(BOARD_SIZE)])
        edge = "     " + "- " * BOARD_SIZE
        
        # Fixed alignment: Header and edge now align with the grid symbols
        print(f"{header}        {header}")
        print(f"{edge}       {edge}")
        
        for y in range(BOARD_SIZE):
            player_row = []
            for x in range(BOARD_SIZE):
                if (x, y) in self.player_board.sunk_cells:
                    player_row.append(self.player_board.get_ship_symbol(x, y).lower())
                elif (x, y) in self.player_board.hits_received:
                    player_row.append("*")  # Hit on active ship
                elif self.player_board.grid[x][y] == 1:
                    player_row.append(self.player_board.get_ship_symbol(x, y))  # Your active ship
                elif (x, y) in self.player_board.misses:
                    player_row.append(".")  # Miss
                else:
                    player_row.append("~")  # Water
            
            enemy_row = []
            for x in range(BOARD_SIZE):
                if (x, y) in self.ai_board.sunk_cells:
                    enemy_row.append(self.ai_board.get_ship_symbol(x, y).lower())
                elif (x, y) in self.ai_board.hits_received:
                    enemy_row.append("*")  # Hit on active enemy ship
                elif (x, y) in self.ai_board.misses:
                    enemy_row.append(".")  # Miss
                elif reveal_enemy and self.ai_board.grid[x][y] == 1:
                    enemy_row.append(self.ai_board.get_ship_symbol(x, y)) # Revealed ship
                else:
                    enemy_row.append("~")  # Water
            
            y_str = str(y).ljust(2)
            print(f"{y_str} | {' '.join(player_row)} |      {y_str} | {' '.join(enemy_row)} |")
        
        # Fixed alignment: Bottom edge now aligns with the grid symbols
        print(f"{edge}       {edge}")

        # --- FLEET STATUS SUMMARY ---
        print("\nFLEET STATUS:")
        max_ship_len = max(len(name) for name, _ in SHIPS)
        print(f"{'  YOUR SHIPS':<{max_ship_len + 12}} | {'  ENEMY SHIPS':<{max_ship_len + 12}}")
        
        for (name, _) in SHIPS:
            p_status = "ALIVE" if name in self.player_board.active_ships else "SUNK "
            e_status = "ALIVE" if name in self.ai_board.active_ships else "SUNK "
            
            p_label = f"{name:<{max_ship_len}} [{p_status}]"
            e_label = f"{name:<{max_ship_len}} [{e_status}]"
            print(f" {p_label}     |  {e_label}")
        print("\nLegend: C/B/S/D/P=Ship, *=Hit, c/b/s/d/p=SUNK, .=Miss, ~=Water")

    def get_player_input(self, active_ships):
        """Prompts the player for ship choice and target coordinates."""
        while True:
            print("\n--- YOUR TURN ---")
            print("Available Ships (Weapons):")
            for i, ship in enumerate(active_ships):
                print(f" {i}: {ship.ljust(12)} Pattern: {len(SHOT_PATTERNS[ship])} squares")
            print(" ?: Show visual pattern guide")
            
            choice = input(f"\nSelect ship to fire with (0-{len(active_ships)-1} or ?): ")
            
            if choice == "?":
                clear_screen()
                show_patterns()
                input("\nPress Enter to return to the game...")
                clear_screen()
                self.render_board()
                continue
                
            try:
                idx = int(choice)
                if 0 <= idx < len(active_ships):
                    ship_name = active_ships[idx]
                    break
                else:
                    print(f"Please enter a number between 0 and {len(active_ships)-1}.")
            except ValueError:
                print("Invalid input. Please enter a number or '?'.")

        target = None
        while target is None:
            try:
                coords = input("Enter target coordinates (x,y) e.g., 5,3: ")
                if "," not in coords:
                    print("Invalid format. Use x,y")
                    continue
                x, y = map(int, coords.split(","))
                if 0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE:
                    target = (x, y)
                else:
                    print(f"Coordinates must be between 0 and {BOARD_SIZE-1}.")
            except ValueError:
                print("Invalid input. Use numbers for x and y.")
                
        return ship_name, target

    def play(self):
        print("Welcome to Tactical Battleship!")
        print("Each ship has a unique shot pattern. Protect your fleet to keep your options open!")
        input("\nPress Enter to begin...")

        while True:
            clear_screen()
            self.render_board()
            
            # --- PHASE 1: PRE-ROUND STATE ---
            # Capture available ships before any shots are fired
            player_active_start = list(self.player_board.active_ships.keys())
            ai_active_start = list(self.ai_board.active_ships.keys())

            # Check for existing game over (from previous round resolution)
            player_dead = len(player_active_start) == 0
            ai_dead = len(ai_active_start) == 0

            if player_dead or ai_dead:
                clear_screen()
                self.render_board(reveal_enemy=True)
                if player_dead and ai_dead:
                    print("\n!!! MUTUAL DESTRUCTION !!!")
                    print("Both fleets have been destroyed. It's a draw!")
                elif ai_dead:
                    print("\n!!! VICTORY !!!")
                    print("You have sunk the entire enemy fleet.")
                elif player_dead:
                    print("\n!!! DEFEAT !!!")
                    print("Your fleet has been destroyed.")
                break
                
            # --- PHASE 2: PLAYER SELECTION ---
            active_ships_sorted = sorted(player_active_start)
            ship_name, (tx, ty) = self.get_player_input(active_ships_sorted)
            pattern = SHOT_PATTERNS[ship_name]
            
            # --- PHASE 3: AI SELECTION (Simultaneous) ---
            # AI selects move based on pre-shot state
            atx, aty, ai_weapon = self.ai_agent.select_move(self.player_board, ai_active_start)
            ai_pattern = SHOT_PATTERNS[ai_weapon]

            # --- PHASE 4: SIMULTANEOUS RESOLUTION ---
            hits = self.ai_board.fire(tx, ty, pattern)
            ai_hits = self.player_board.fire(atx, aty, ai_pattern)
            
            clear_screen()
            self.render_board()
            
            # Show Player Result
            print(f"\nYOU FIRED: {ship_name} at ({tx},{ty}).")
            if hits > 0:
                print(f"BOOM! {hits} hit(s) recorded!")
            else:
                print("Splash... You missed.")
            
            # Show AI Result
            print(f"\nAI FIRED: {ai_weapon} at ({atx},{aty}).")
            if ai_hits > 0:
                print(f"DANGER! The enemy scored {ai_hits} hit(s) on your fleet!")
            else:
                print("The enemy missed.")
                
            input("\nPress Enter for next round...")

def main():
    game = Game()
    game.play()

if __name__ == "__main__":
    main()

File: game.ts
import { createInterface } from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';
import { AI, placeFleetRandomly } from './src/ai';
import { Board, SHIP_NAMES, BattleshipMatch, SHOT_PATTERNS } from './src/engine';
import { mulberry32 } from './src/ai-utils';
import {
  COLUMN_LABELS,
  PROMPT_SHIP_ORDER,
  DISPLAY_SHIP_SYMBOLS,
  VALID_VARIANTS,
  isVariant,
  type AIVariants,
  clearScreen,
  formatCoord,
  parseCoord,
  renderBoard,
  renderHeatmap,
  printPatternHelp,
  parseVars,
  validateVars,
  parseCommonArgs,
} from './src/cli';

const renderInteractiveBoards = (
  playerBoard: Board,
  aiBoard: Board,
  aiController: AI,
  showHeatmap: boolean,
  statusLine = '',
  pendingMove?: { origin: { x: number; y: number }; impactCells: Set<string> },
) => {
  clearScreen();

  if (showHeatmap) {
    console.log(
      renderHeatmap(
        'Enemy Waters',
        aiBoard,
        aiController.getHeatmap(aiBoard).heatmap,
        false,
        pendingMove,
      ),
    );
    console.log();
    console.log(renderHeatmap('Your Fleet', playerBoard, aiController.getHeatmap(playerBoard).heatmap));
  } else {
    console.log(renderBoard('Enemy Waters', aiBoard, false, pendingMove));
    console.log();
    console.log(renderBoard('Your Fleet', playerBoard, true));
  }

  console.log(`\n${statusLine}`);
};

const runWatchMode = async (
  sideAVariant: AIVariants,
  sideBVariant: AIVariants,
  effectiveFleetSeed: number | undefined,
  effectiveMoveSeed: number | undefined,
  mirror: boolean,
  auto: boolean,
  rl: ReturnType<typeof createInterface>,
  configA?: Record<string, number | boolean>,
  configB?: Record<string, number | boolean>,
) => {
  // Seeding logic matches sim.ts exactly:
  // without mirror: both sides share the same RNG instance (A draws first, B draws next)
  // with mirror: each side gets an independent instance from the same seed (identical sequences)
  const rngFleet = effectiveFleetSeed !== undefined ? mulberry32(effectiveFleetSeed) : undefined;
  const rngFleetA = rngFleet;
  const rngFleetB = mirror && effectiveFleetSeed !== undefined ? mulberry32(effectiveFleetSeed) : rngFleet;
  const rngMoveA = effectiveMoveSeed !== undefined ? mulberry32(effectiveMoveSeed) : undefined;
  const rngMoveB = mirror && effectiveMoveSeed !== undefined ? mulberry32(effectiveMoveSeed) : rngMoveA;

  const aiA = new AI(sideAVariant, configA);
  const aiB = new AI(sideBVariant, configB);
  const fleetA = aiA.placeFleet(undefined, rngFleetA);
  const fleetB = aiB.placeFleet(undefined, rngFleetB);
  const match = new BattleshipMatch(fleetA, fleetB, 'sideA', 'sideB');

  const fleetSeedStr = effectiveFleetSeed !== undefined ? String(effectiveFleetSeed) : '(random)';
  const moveSeedStr = effectiveMoveSeed !== undefined ? String(effectiveMoveSeed) : '(random)';
  console.log(`Watch: ${sideAVariant} vs ${sideBVariant}${mirror ? ' [MIRRORED]' : ''}`);
  console.log(`fleet-seed: ${fleetSeedStr}  move-seed: ${moveSeedStr}\n`);
  console.log(renderBoard(`Side A (${sideAVariant})`, match.boardA, true));
  console.log();
  console.log(renderBoard(`Side B (${sideBVariant})`, match.boardB, true));

  if (!auto) await rl.question('\nPress Enter to start...');

  let showHeatmap = false;
  let totalHitsA = 0;
  let totalHitsB = 0;
  let quit = false;

  while (!match.isGameOver && !quit) {
    const moveA = aiA.selectMove(match.boardB, match.boardA.getActiveShipNames(), rngMoveA);
    const moveB = aiB.selectMove(match.boardA, match.boardB.getActiveShipNames(), rngMoveB);

    let result: any = null;

    const renderState = (phase: 1 | 2) => {
      clearScreen();
      const round = match.round + (phase === 2 ? 0 : 1);
      const avgA = (totalHitsA / Math.max(1, match.round)).toFixed(2);
      const avgB = (totalHitsB / Math.max(1, match.round)).toFixed(2);

      const aliveA = match.boardA.getActiveShipNames().join(' ') || 'none';
      const aliveB = match.boardB.getActiveShipNames().join(' ') || 'none';
      if (showHeatmap) {
        const previewA =
          phase === 1
            ? {
                origin: { x: moveB.x, y: moveB.y },
                impactCells: new Set(
                  match.boardA.getImpactCoordinates(moveB.x, moveB.y, SHOT_PATTERNS[moveB.ship] ?? []),
                ),
              }
            : undefined;
        const previewB =
          phase === 1
            ? {
                origin: { x: moveA.x, y: moveA.y },
                impactCells: new Set(
                  match.boardB.getImpactCoordinates(moveA.x, moveA.y, SHOT_PATTERNS[moveA.ship] ?? []),
                ),
              }
            : undefined;

        // A's board overlaid with B's targeting heatmap (B is attacking A)
        console.log(
          renderHeatmap(
            `Side A (${sideAVariant}) [alive: ${aliveA}]`,
            match.boardA,
            aiB.getHeatmap(match.boardA).heatmap,
            false,
            previewA,
          ),
        );
        console.log();
        // B's board overlaid with A's targeting heatmap (A is attacking B)
        console.log(
          renderHeatmap(
            `Side B (${sideBVariant}) [alive: ${aliveB}]`,
            match.boardB,
            aiA.getHeatmap(match.boardB).heatmap,
            false,
            previewB,
          ),
        );
        console.log('\nLegend: 0-9 scaled heat, $ = 100%, @ = target');
      } else {
        const previewA =
          phase === 1
            ? {
                origin: { x: moveB.x, y: moveB.y },
                impactCells: new Set(
                  match.boardA.getImpactCoordinates(moveB.x, moveB.y, SHOT_PATTERNS[moveB.ship] ?? []),
                ),
              }
            : undefined;
        const previewB =
          phase === 1
            ? {
                origin: { x: moveA.x, y: moveA.y },
                impactCells: new Set(
                  match.boardB.getImpactCoordinates(moveA.x, moveA.y, SHOT_PATTERNS[moveA.ship] ?? []),
                ),
              }
            : undefined;

        console.log(renderBoard(`Side A (${sideAVariant}) [alive: ${aliveA}]`, match.boardA, true, previewA));
        console.log();
        console.log(renderBoard(`Side B (${sideBVariant}) [alive: ${aliveB}]`, match.boardB, true, previewB));
      }
      console.log(`\nRound ${round} | A avg hits/round: ${avgA} | B avg hits/round: ${avgB}`);
      
      if (phase === 1) {
        console.log(`→ Side A (${moveA.ship}) targets ${formatCoord(moveA.x, moveA.y)}  |  Side B (${moveB.ship}) targets ${formatCoord(moveB.x, moveB.y)}`);
      } else {
        const hitsAStr = result.attackA.hits > 0 ? `+${result.attackA.hits} hit${result.attackA.sunkShips.length ? `, sunk: ${result.attackA.sunkShips.join(', ')}` : ''}` : 'miss';
        const hitsBStr = result.attackB.hits > 0 ? `+${result.attackB.hits} hit${result.attackB.sunkShips.length ? `, sunk: ${result.attackB.sunkShips.join(', ')}` : ''}` : 'miss';
        console.log(`A: ${moveA.ship} → ${formatCoord(moveA.x, moveA.y)} (${hitsAStr})`);
        console.log(`B: ${moveB.ship} → ${formatCoord(moveB.x, moveB.y)} (${hitsBStr})`);
        if (match.isGameOver) {
          const winner = match.winner;
          if (winner === 'draw') console.log(`DRAW: Both fleets destroyed after ${match.round} rounds.`);
          else if (winner === 'sideA') console.log(`WINNER: Side A (${sideAVariant}) after ${match.round} rounds.`);
          else console.log(`WINNER: Side B (${sideBVariant}) after ${match.round} rounds.`);
        }
      }
    };

    renderState(1);

    if (!auto) {
      let advancing = false;
      while (!advancing) {
        const key = (await rl.question('\nEnter=fire, h=heatmap, q=quit: ')).trim().toLowerCase();
        if (key === 'q') { quit = true; advancing = true; }
        else if (key === 'h') { showHeatmap = !showHeatmap; renderState(1); }
        else { advancing = true; }
      }
      if (quit) break;
    }

    result = match.resolveTurn(moveA, moveB);
    totalHitsA += result.attackA.hits;
    totalHitsB += result.attackB.hits;

    renderState(2);

    if (!auto && !match.isGameOver) {
      let advancing = false;
      while (!advancing) {
        const key = (await rl.question('\nEnter=next turn, h=heatmap, q=quit: ')).trim().toLowerCase();
        if (key === 'q') { quit = true; advancing = true; }
        else if (key === 'h') { showHeatmap = !showHeatmap; renderState(2); }
        else { advancing = true; }
      }
    }
  }
};

const getPlayerMove = async (
  rl: ReturnType<typeof createInterface>,
  playerBoard: Board,
  targetBoard: Board,
  activeShips: string[],
  aiController: AI,
) => {
  let showHeatmap = false;
  let statusLine = '';
  let pending: { ship: string; x: number; y: number } | null = null;

  while (true) {
    const impactCells = pending
      ? new Set(targetBoard.getImpactCoordinates(pending.x, pending.y, SHOT_PATTERNS[pending.ship] ?? []))
      : new Set<string>();

    renderInteractiveBoards(
      playerBoard,
      targetBoard,
      aiController,
      showHeatmap,
      statusLine,
      pending ? { origin: { x: pending.x, y: pending.y }, impactCells } : undefined,
    );

    const shipList = PROMPT_SHIP_ORDER.filter((s) => activeShips.includes(s))
      .map((s) => `${PROMPT_SHIP_ORDER.indexOf(s) + 1}:${s}`)
      .join(', ');

    const prompt = pending
      ? `Previewing ${pending.ship} at ${formatCoord(pending.x, pending.y)}. Enter to FIRE, or new command: `
      : `Enter command (e.g. '1 A5' or 'Carrier A5', ships: ${shipList}, h, ?): `;

    const inputRaw = (await rl.question(`\n${prompt}`)).trim();

    if (inputRaw === '' && pending) {
      return pending;
    }

    if (inputRaw === '?') {
      printPatternHelp();
      await rl.question('\nPress Enter to continue...');
      continue;
    }

    if (inputRaw.toLowerCase() === 'h') {
      showHeatmap = !showHeatmap;
      statusLine = showHeatmap ? 'Heat map view enabled.' : '';
      continue;
    }

    const parts = inputRaw.split(/\s+/);
    if (parts.length < 2) {
      statusLine = 'Invalid command format. Use "[ship] [coordinate]", e.g., "1 A5".';
      pending = null;
      continue;
    }

    const [shipToken, coordToken] = parts;
    const selectedShip = PROMPT_SHIP_ORDER[Number(shipToken) - 1];
    const matchedShip =
      (selectedShip && activeShips.includes(selectedShip) ? selectedShip : null) ??
      activeShips.find((s) => s.toLowerCase() === shipToken.toLowerCase());

    if (!matchedShip) {
      statusLine = `Invalid ship: ${shipToken}.`;
      pending = null;
      continue;
    }

    const parsedCoord = parseCoord(coordToken);
    if (!parsedCoord) {
      statusLine = `Invalid coordinate: ${coordToken}.`;
      pending = null;
      continue;
    }

    if (!targetBoard.canTargetCell(parsedCoord.x, parsedCoord.y, matchedShip)) {
      statusLine = 'That shot pattern cannot hit any new cell from that origin.';
      pending = null;
      continue;
    }

    pending = { x: parsedCoord.x, y: parsedCoord.y, ship: matchedShip };
    statusLine = '';
  }
};

const main = async () => {
  const options = parseCommonArgs(process.argv.slice(2));
  const { ai, help, watch, sideA, sideB, fleetSeed, moveSeed, seed, game, mirror, auto, vars, varsA, varsB } = options;

  const watchFlags = ['--sideA', '--sideB', '--varsA', '--varsB', '--fleet-seed', '--move-seed', '--seed', '--game', '--mirror', '--auto'];
  const watchFlagsUsed = process.argv.some((arg) => watchFlags.includes(arg));

  if (help) {
    console.log([
      'Usage:',
      `  Normal game:  npx tsx game.ts [--ai <variant>] [--vars key=val,...]`,
      `  Watch mode:   npx tsx game.ts --watch [--sideA <variant>] [--sideB <variant>]`,
      `                  [--varsA key=val,...] [--varsB key=val,...]`,
      `                  [--fleet-seed <n>] [--move-seed <n>]`,
      `                  [--seed <n>] [--game <i>]  (fleet-seed=seed+i*2, move-seed=seed+i*2+1)`,
      `                  [--mirror] [--auto]`,
      '',
      `Variants: ${VALID_VARIANTS.join(', ')}`,
    ].join('\n'));
    return;
  }

  if (watchFlagsUsed && !watch) {
    console.error('Error: --sideA, --sideB, --varsA, --varsB, --fleet-seed, --move-seed, --seed, --game, --mirror, and --auto require --watch.');
    process.exit(1);
  }

  if (watch && Object.keys(vars).length > 0) {
    console.error('Error: --vars is not allowed in --watch mode. Use --varsA and --varsB instead.');
    process.exit(1);
  }

  if (watch) {
    if (!isVariant(sideA)) {
      console.error(`Invalid --sideA variant. Use one of: ${VALID_VARIANTS.join(', ')}.`);
      process.exit(1);
    }
    if (!isVariant(sideB)) {
      console.error(`Invalid --sideB variant. Use one of: ${VALID_VARIANTS.join(', ')}.`);
      process.exit(1);
    }

    const parsedVarsA = validateVars(varsA, sideA, '--varsA');
    if (parsedVarsA === null) process.exit(1);
    const parsedVarsB = validateVars(varsB, sideB, '--varsB');
    if (parsedVarsB === null) process.exit(1);

    if (game !== undefined && seed === undefined) {
      console.error('Error: --game requires --seed');
      process.exit(1);
    }

    const gameIdx = game ?? 0;
    const effectiveFleetSeed = fleetSeed ?? (seed !== undefined ? seed + gameIdx * 2 : undefined);
    const effectiveMoveSeed = moveSeed ?? (seed !== undefined ? seed + gameIdx * 2 + 1 : undefined);

    const rl = createInterface({ input, output });
    try {
      await runWatchMode(sideA, sideB, effectiveFleetSeed, effectiveMoveSeed, mirror, auto, rl, parsedVarsA, parsedVarsB);
    } finally {
      rl.close();
    }
    return;
  }

  if (!isVariant(ai)) {
    console.error(`Invalid AI variant. Use one of: ${VALID_VARIANTS.join(', ')}.`);
    process.exit(1);
  }

  const parsedVars = validateVars(vars, ai, '--vars');
  if (parsedVars === null) process.exit(1);

  const rl = createInterface({ input, output });
  const aiController = new AI(ai, parsedVars);
  const match = new BattleshipMatch(placeFleetRandomly(), aiController.placeFleet(), 'player', 'ai');

  try {
    console.log('Welcome to Tactical Battleship!');
    console.log(`AI variant: ${ai}`);
    await rl.question('\nPress Enter to begin...');

    while (true) {
      const playerDead = match.boardA.isGameOver();
      const aiDead = match.boardB.isGameOver();

      if (playerDead || aiDead) {
        clearScreen();
        console.log(renderBoard('Enemy Waters', match.boardB, true));
        console.log();
        console.log(renderBoard('Your Fleet', match.boardA, true));

        if (playerDead && aiDead) console.log("\n!!! MUTUAL DESTRUCTION !!!\nBoth fleets have been destroyed. It's a draw!");
        else if (aiDead) console.log('\n!!! VICTORY !!!\nYou have sunk the entire enemy fleet.');
        else console.log('\n!!! DEFEAT !!!\nYour fleet has been destroyed.');
        break;
      }

      const playerMove = await getPlayerMove(
        rl,
        match.boardA,
        match.boardB,
        match.boardA.getActiveShipNames(),
        aiController,
      );
      const aiMove = aiController.selectMove(match.boardA, match.boardB.getActiveShipNames());

      const result = match.resolveTurn(playerMove, aiMove);

      clearScreen();
      console.log(renderBoard('Enemy Waters', match.boardB, false));
      console.log();
      console.log(renderBoard('Your Fleet', match.boardA, true));
      console.log(`\nYOU FIRED: ${playerMove.ship} at ${formatCoord(playerMove.x, playerMove.y)}.`);
      console.log(result.attackA.hits > 0 ? `BOOM! ${result.attackA.hits} hit(s) recorded!` : 'Splash... You missed.');
      result.attackA.sunkShips.forEach((ship) => console.log(`Enemy ${ship} destroyed!`));
      console.log(`\nAI FIRED: ${aiMove.ship} at ${formatCoord(aiMove.x, aiMove.y)}.`);
      console.log(result.attackB.hits > 0 ? `DANGER! The enemy scored ${result.attackB.hits} hit(s) on your fleet!` : 'The enemy missed.');
      result.attackB.sunkShips.forEach((ship) => console.log(`CRITICAL: Your ${ship} was destroyed!`));

      if (result.winnerId) {
        await rl.question('\nPress Enter to view final result...');
      } else {
        await rl.question('\nPress Enter for next round...');
      }
    }
  } finally {
    rl.close();
  }
};

main();
